components/places/src/storage/history/mutations.rs

```diff
@@ -0,0 +1,491 @@
+            db.execute_cached(
+                &format!(
+                    "
+                    INSERT OR IGNORE INTO moz_historyvisit_tombstones(place_id, visit_date)
+                    SELECT place_id, visit_date
+                    FROM moz_historyvisits
+                    WHERE id IN ({})
+                    ",
+                    var_repeat,
+                ),
+                params.clone(),
+            )?;
+
+            db.execute_cached(
+                &format!("DELETE FROM moz_historyvisits WHERE id IN ({})", var_repeat),
+                params,
+            )?;
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    fn recalc_pages(db: &PlacesDb, page_ids: HashSet<RowId>) -> Result<()> {
+        let mut pages_to_clean: Vec<PageToClean> = vec![];
+        sql_support::each_chunk(&Vec::from_iter(page_ids), |chunk, _| -> Result<()> {
+            pages_to_clean.append(&mut db.query_rows_and_then_cached(
+                &format!(
+                    "SELECT
+                    id,
+                    (foreign_count != 0) AS has_foreign,
+                    ((last_visit_date_local + last_visit_date_remote) != 0) as has_visits,
+                    sync_status
+                FROM moz_places
+                WHERE id IN ({})",
+                    sql_support::repeat_sql_vars(chunk.len())
+                ),
+                rusqlite::params_from_iter(chunk),
+                PageToClean::from_row,
+            )?);
+            Ok(())
+        })?;
+        cleanup_pages(db, &pages_to_clean)?;
+        Ok(())
+    }
+
+    fn delete_from_pending_temp_tables(db: &PlacesDb) -> Result<()> {
+        crate::storage::delete_pending_temp_tables(db)
+    }
+}
+
+/// Stores a visit that we want to delete.  These can be build from queries that include the `id`
+/// and `place_id` columns from `moz_historyvisits`
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(super) struct VisitToDelete {
+    pub(super) visit_id: RowId,
+    pub(super) page_id: RowId,
+}
+
+impl VisitToDelete {
+    pub(super) fn from_row(row: &Row<'_>) -> Result<Self> {
+        Ok(Self {
+            visit_id: row.get("id")?,
+            page_id: row.get("place_id")?,
+        })
+    }
+}
+
+/// Create a Vec of mutations to delete visits returned by a query
+///
+/// The query must include the id and place_id columns from moz_historyvisits.
+pub(super) fn delete_visit_mutations(
+    visits_to_delete: impl IntoIterator<Item = VisitToDelete>,
+) -> Vec<DbMutation> {
+    let mut visit_ids = HashSet::<RowId>::new();
+    let mut page_ids = HashSet::<RowId>::new();
+    for visit_to_delete in visits_to_delete.into_iter() {
+        visit_ids.insert(visit_to_delete.visit_id);
+        page_ids.insert(visit_to_delete.page_id);
+    }
+    vec![
+        DbMutation::DeleteVisitRows { visit_ids },
+        DbMutation::RecalcPages { page_ids },
+        DbMutation::DeleteFromPendingTempTables,
+    ]
```

Comment:
I'm pretty sure these are the correct steps when deleting visits from the DB, but please double check that I'm not missing anything.

Reply:
That looks correct to me!